### ISSUE TITLE
Zooming the map by scrolling does not work as expected on Touchpad

### ISSUE THREAD

**Issue by afontenot** [mrkite/minutor#224]
Hi,

I have a touchpad that uses two finger inputs to scroll. Scrolling works in every other program. In Minutor, attempting to scroll will either do nothing, zoom *in* to a random zoom level, or zoom *out* to a random zoom level (including randomly jerking in and out across multiple zoom levels even if I am scrolling continuously on one direction on the touchpad). 

This is on KDE Plasma (latest released version) on Linux. I believe my touchpad is using the xf86-input-synaptics driver.

Things I have tried:

 * Set up a script with `xdotool` to send one "scroll up" input every five seconds. This *works*. Minutor goes up one zoom level every five seconds.

 * Checked the inputs being sent to programs by the mouse driver with `xev`. If I scroll extremely slowly, I can get the touchpad to only send two scroll events at a time. I can't get it down to one. Note that scrolling extremely slowly does not seem to solve the problem with scrolling in Minutor.

If you don't immediately know how to fix this problem, or fixing it would create other issues, I think it would be nice to have some kind of keyboard input for scrolling as a around, and possible a zoom slider on the right of the screen as well, like the "layers" slider.

**Comment by EtlamGit** [comment]
Mouse scroll wheel is ZOOM, so the behavior is intended.

Keyboard navigation is available. Just use the Cursor Keys.
You can also use modifier keys (like SHIFT ALT CTRL) to align to Chunks and whatever.

**Comment by afontenot** [comment]
> Mouse scroll wheel is ZOOM, so the behavior is intended.

Can you explain what you mean by "intended"? I understand that scroll is supposed to be zoom, but it's not functional *as* a zoom control for me, as I described in the issue.

> Keyboard navigation is available. Just use the Cursor Keys.

Sorry, could you explain what "cursor keys" are? A search suggests you might be referring to arrow keys or the numpad cursor control keys (when enabled). Both of these *pan* the map for me, not zoom it. This issue is about the zoom function being broken for me, I'm going to edit the title to make that more clear.

**Comment by afontenot** [comment]
Here is a short screencast of the problem. Although my mouse inputs aren't clearly visible, what I'm attempting to do is slowly zoom in using my touchpad. It doesn't work at all most of the time, then suddenly zooms in about 10x or so.

https://drive.google.com/file/d/1raQfAoZx0MDxzbtMiorZMuPkU4SlF35O/view?usp=sharing

**Comment by EtlamGit** [comment]
Maybe I just misunderstood your explanation...
A next try with closer reading:
- You have a computer with a touchpad. Most probably a laptop and not a touch-only device like a assumed before.
- You use some sort of Linux operation system with KDE desktop
- You use two-finger-gestures to scroll. This should be two fingers next to each other and moved up or down on the touchpad.
- Somewhere in your software or configuration multiple scroll events are generated for just tiny movement of your fingers.

As far as I remember, I solved a similar issue some years ago. Therefore a question, do you use latest version of Minutor? Did you try with a self compiled version? (Latest release is very old).

There should be a Qt value we can query about the expected events per scroll wheel turn. I will look into that direction (no time immediately, more in the next weeks).

**Comment by afontenot** [comment]
I'm compiling it, the one I was using when I opened this was https://github.com/mrkite/minutor/commit/57a47ff which is from the end of March. I'm trying to update to make sure the issue wasn't fixed, but I can't compile it at present. Bisecting reveals that https://github.com/mrkite/minutor/commit/c456e1d75dfc60a951d72d107616ec1ebffb6b62 introduced the problem. (Yes, I cherrypicked https://github.com/mrkite/minutor/commit/5d5a65ee753dcc182bce38338a42086915a891b1 to avoid that compilation issue.)

    /usr/bin/ld: properties.o: in function `Properties::DisplayProperties(QVariant)':
    properties.cpp:(.text+0xe8): undefined reference to `void PropertieTreeCreator::ParseIterable<QHash<QString, QVariant> >(QTreeWidgetItem*, QHash<QString, QVariant> const&)'
    /usr/bin/ld: properties.cpp:(.text+0x1e7): undefined reference to `void PropertieTreeCreator::ParseIterable<QMap<QString, QVariant> >(QTreeWidgetItem*, QMap<QString, QVariant> const&)'

I'm on QT version 5.15.0. I can copy this into a new issue if you want.

I think it's worth noting that most of the time Minutor doesn't seem to react to my scroll events at all. The bug would be more understandable to me if it always zoomed in more quickly than I wanted because it was reacting to 2 scroll events instead of just one, but most of the time it seems to ignore them.

**Comment by mctynp** [comment]
The QTWheelEvent delta() function is deprecated according to the current QT 5 docs. angleDelta() should be supported on all platforms, with pixelDelta() for some platforms (macOS is called out in the docs, with a mention that pixelDelta() is unreliable and driver-dependent on X11). Replacing delta() with angleDelta() seems to work for zooming with a mouse wheel under Windows in a quick local test, but I don't know how it will perform under other operating systems, or with a touchpad.

If that's not a viable fix, it looks like the page up/page down keys allow for keyboard zooming.

**Comment by magicus** [comment]
FWIW, I have the same issue on my MacBook: I cannot zoom. The "standard" mac touchpad gesture of pinching is not supported, and the "emulate scroll wheel by dragging two fingers" exhibit just the kind of jerky, random (and mostly non-working) behavior as @afontenot describes in the bug report.

**Comment by rjp** [comment]
Can confirm that the latest version of Minutor (2.18.0) still suffers from this problem on macOS Monterey (12.1).  Seems to activate only if I do a harsh flick which then instantly zooms in or out to the max/min.

(I can, however, zoom reliably using FN-up/down cursors which makes it less of an issue for me.)

**Comment by bighairydave** [comment]
I can confirm the issue is as described by OP on an M1 MacBook Pro with macOS Monterey.

I have also found scroll-wheel input to be at best unpredictable with a Bluetooth-connected Razer mouse. Scrolling action only occasionally changes the view. Tilting the wheel (which is supposed to be an accelerated zoom) will zoom to max/min views.

I learned from the above comment that holding fn and using up and down cursor keys will zoom in and out. Also fn-left and fn-right will move the depth slider. This may be worth documenting in the README, especially if it holds true for Linux laptops.

**Comment by afontenot** [comment]
Finally got around to fixing this myself. There might be multiple issues here, so testing (especially from anyone on macOS) would be appreciated. Note that until this is merged, you'll have to build from my tree.

**Comment by bighairydave** [comment]
I've tested the new builds on MacOS and this fixes the zoom issue. Zooming now works as expected with either touchpad or mouse wheel.

Could the same fix be implemented for the depth slider?

**Comment by afontenot** [comment]
@bighairydave Could you clarify the issue you're seeing with the depth slider and file a separate issue for that if necessary? I don't have any problems with it (other than redraws being a little slow).

**Comment by EtlamGit** [comment]
(other than redraws being a little slow)
-> a redraw after Depth Slider change has to process & render all data again, so it has to be slower than just panning.

**Comment by bighairydave** [comment]
Yeah, I didn't explain this well. The depth slider is fine, and redrawing is pretty much instant on this machine. The issue is that scrolling while holding shift doesn't move the depth slider as expected. I'm willing to bet that the fix is similar to the one for the zoom issue!

**Comment by EtlamGit** [comment]
As I do not have any i-devices, could you all please describe what devices you are using exactly?
Maybe with link to a Wikipedia article about it.
(I even do not know whether you are talking about a phone or a computer)

And for those that are able to compile them self:
Please debug into `MapView::wheelEvent(QWheelEvent *event)`
and provide data of `event` when you are using your touchpad/mouse/whatever
and return values for:
* `QWheelEvent::pixelData()`
* `QWheelEvent::source()`
